Add HTTP Headers to Webhook Agent

Fishwaldo opened the pull request:
This adds the received HTTP_* headers to the Webhook Agent. This is useful for services such as GitHub that sent the Webhook Type as a header.

As Per issue #2431

dsander commented on app/models/agents/webhook_agent.rb:
```diff
@@ -84,7 +92,8 @@ def receive_web_request(params, method, format)
           return ["Not Authorized", 401]
       end
-
+      params[interpolated['header_key']] = header_for(headers)
```
I think we should check if `interpolated['header_key']` is present and only apply the headers in that case. Otherwise we would write them as `nil` in `params`.

Fishwaldo replied:
Ok. I actually moved it around a bit. We wont add the headers if they header_key is empty, but I've also moved the logic inside the payload_for loop so the headers are present regardless of payload_path.

Fishwaldo commented:
Updated the spec file. I noticed some of the recaptcha tests are failing. But it appears to be on the Google side (at least for me) and not in the agent. 

Also - Updated the headers_path to be just a comma seperated list of headers the user wants, as I guess they could want more than 1!

dsander commented on spec/models/agents/webhook_agent_spec.rb:
```diff
@@ -321,13 +521,89 @@
+        out = nil
+
+        expect {
+          out= agent.receive_web_request(webpayload)
```
Missing space after `out` 😄

dsander commented on spec/models/agents/webhook_agent_spec.rb:
```diff
@@ -284,33 +471,46 @@
           out = nil
 
           stub_request(:any, /verify/).to_return { |request|
+            puts gotrequest
```
✂️ please

dsander commented on spec/models/agents/webhook_agent_spec.rb:
```diff
@@ -284,33 +471,46 @@
+            puts gotrequest
             checked = true
-            { status: 200, body: '{"success":false}' }
+           { status: 200, body: '{"success":false}' }
```
I think that indentation should stay the way it was

dsander commented on app/models/agents/webhook_agent.rb:
```diff
@@ -80,15 +86,20 @@ def receive_web_request(params, method, format)
           return ["Not Authorized", 401]
         end
-
+          
```
✂️ the trailing white space please

dsander commented on app/models/agents/webhook_agent.rb:
```diff
@@ -80,15 +86,20 @@ def receive_web_request(params, method, format)
         create_event(payload: payload)
       end
-
+      
```
✂️ same here

dsander commented on app/models/agents/webhook_agent.rb:
```diff
@@ -80,15 +86,20 @@ def receive_web_request(params, method, format)
+        
 
       [payload_for(params)].flatten.each do |payload|
+        if interpolated['header_key'].presence 
```
`presence` also returns the value, since we don't need it here `present?` is enough

dsander commented on app/models/agents/webhook_agent.rb:
```diff
@@ -80,15 +86,20 @@ def receive_web_request(params, method, format)
 
       [payload_for(params)].flatten.each do |payload|
+        if interpolated['header_key'].presence 
+          acceptedheaders = (interpolated['headers']).split(/,/).map { |x| x.strip }.select
```
I don't think we need to quotes around `interpolated['headers']` and the trailing `select`

Fishwaldo commented:
I’m going to have a new PR soon to add a option to whitelist IP addresses that can make requests as well. Just FYI.

degan6 commented:
Can't wait to see this feature added!

Fishwaldo commented:
Sorry for the fileperm issues. Dealing with a funky install of Ubuntu on Windows (via WSL).

knu commented:
@Fishwaldo Thank you for your contribution!
